Free all memory allocated by LKL (except for undetached pthreads)

pscollins opened the pull request:
This commit undoes the issue introduced when we had to stop running `free_mem` at shutdown, and gives a nearly-clean Valgrind output. In conjunction with #127, it gives totally clean Valgrind output except for a few false positives in `dlerror` and `timer_create`. It depends on #125, which I submitted separately so that it wouldn't be gated on all of the changes in this PR --- I can either close that or rebase this one on top of master once it's resolved, whichever is easier.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/lkl/linux/130)

<!-- Reviewable:end -->

liuyuan10 commented on tools/lkl/lib/virtio_net_tap.c:
```diff
@@ -122,5 +196,24 @@ struct lkl_netdev *lkl_netdev_tap_create(const char *ifname)
 		return NULL;
 	}
 
+	#ifdef CONFIG_EVENTFD
```
Using CONFIG_EVENTFD seems very hacky. What if I want lkl_sys_eventfd and to close the virtio? We are calling all kinds of host sys calls in this file. Doesn't it look fragile?

pscollins replied:
It is really hacky, but it was the best I could think of. The problem is
that if eventfd has been hijacked then we get an LKL fd out of the call and
we can't poll on it in the same fd set as the host fd attached to the TAP
device. And since there's no way to signal to eventfd that we want to hit
the host, we would need to dlsym in a pointer to the host version.

There's also no easy way to tell if we've been hijacked -- we could set
some variable in a hijack constructor function and check it in here, but
then the hijack logic is spread around the whole codebase. I figured we
would want to keep the hijack logic out of the main codebase because we
won't always want to use it.

We do depend heavily on host functions here, but TAP is POSIX only so it's
safe. Also hiding these calls behind the host_ops stuff doesn't prevent it
from being hijacked.

What do you think is the best thing to do? We could add another CONFIG
option to turn on this behavior.
On Mar 16, 2016 6:26 PM, "Yuan Liu" <notifications@github.com> wrote:

> In tools/lkl/lib/virtio_net_tap.c
> <https://github.com/lkl/linux/pull/130#discussion_r56443682>:
>
> > @@ -122,5 +196,24 @@ struct lkl_netdev *lkl_netdev_tap_create(const char *ifname)
> >  		return NULL;
> >  	}
> >
> > +	#ifdef CONFIG_EVENTFD
>
> Using CONFIG_EVENTFD seems very hacky. What if I want lkl_sys_eventfd and
> to close the virtio? We are calling all kinds of host sys calls in this
> file. Doesn't it look fragile?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/lkl/linux/pull/130/files/768b8f7f19c4c8f6c3f9d71531e68ba9d4662a73#r56443682>
>

liuyuan10 replied:
Basically I don't think the code in virtio_net_tap should care about anything of hijack. The hijack lib should make sure it works.

One possible solution is making lkl_netdev_tap_create accept an function pointer which by default is eventfd. This at least make it possible to have eventfd and free up mem safely.

More generally, maybe we should have a struct lkl_tap_host_ops, which defines all the host ops needed by tap. 
struct lkl_tap_host_ops {
  int (*eventfd)(unsigned int, int);
}

It's exposed by virtio_net_tap.h and modified by hijack lib. Dont know if the virtio tap is the only reason we have the hacky 500 fd limit. But this also can help to solve it.

pscollins replied:
I think the 500 fd limit with hijacking also comes up because of stuff like `fprintf(STDOUT, ...)`. It's confined to the hijacking code, though, so it's not too bad.

I understand your solution to be saying something like:

in `virtio_net_tap.{h,c}`:
```
struct lkl_tap_host_ops {
    .eventfd = eventfd, // refers to host eventfd if not hijacked, but LKL eventfd if hijacked
} lkl_tap_host_ops_vals;
```

and somewhere in `hijack/init.c`:

```
#import "virtio_net_tap.h"

__attribute__((constructor)) void fixup_lkl_tap_host_ops_vals(void)
{
    lkl_tap_host_ops_vals.eventfd = dlsym(RTLD_NEXT, "eventfd");
}
```

and "undo" the hijacking. Do I understand correctly?

It looks a little bit awkward to me to say `.eventfd = eventfd` since we'll never need that layer of indirection without hijacking. But it allows for this code to work as expected while `eventfd` is hijacked, and it provides more flexibility in the future if we want to swap other functions into this --- it might make unit testing significantly easier.

I had something in mind to provide `CONFIG_HIJACK_EVENTFD` and `CONFIG_CLEANUP_NETDEVS` as compile-time options and say that they're mutually exclusive, but I think your solution is better. I'll change this commit tomorrow.

liuyuan10 replied:
Yes. Your understanding is correct. Feel free to invent any method. My baseline is we need support using eventfd and freeing up virtio at the same time.

liuyuan10 commented on tools/lkl/lib/virtio_net_tap.c:
```diff
@@ -122,5 +196,24 @@ struct lkl_netdev *lkl_netdev_tap_create(const char *ifname)
+	#else  /* CONFIG_EVENTFD */
+	/* eventfd can't be hijacked, so we can use it */
+	nd->dev.ops = &tap_net_ops_closeable;
+	nd->eventfd = eventfd(0, EFD_NONBLOCK | EFD_SEMAPHORE);
```
What's the reason we can't use close(nd->fd) to stop the polling threads?

liuyuan10 replied:
After looking at the man page, I think it's buggy to use close.
If a file descriptor being monitored by select() is closed in another thread, the result is unspecified. On some UNIX systems, select() unblocks and returns, with an indication that the file descriptor is ready (a subsequent I/O operation will likely fail with an error, unless another the file descriptor reopened between the time select() returned and the I/O operations was performed). On Linux (and some other systems), closing the file descriptor in another thread has no effect on select(). In summary, any application that relies on a particular behavior in this scenario must be considered buggy. 

sorry about this. I should have studied it before shouting out loudly.

pscollins commented:
@liuyuan10 I've made the changes you suggested --- take a look at this version when you have some time and let me know what you think.

opurdila commented on tools/lkl/lib/virtio_net.c:
```diff
@@ -124,6 +128,22 @@ void poll_thread(void *arg)
 }
 
+struct virtio_net_dev *registered_devs[MAX_NET_DEVS];
+static int registered_dev_idx = 0;
```
There is another counter in lkl_netdev_add, I think we can remove it from now and use this one.

pscollins replied:
Done

opurdila commented on tools/lkl/lib/virtio_net.c:
```diff
@@ -238,3 +238,42 @@ int lkl_netdev_add(struct lkl_netdev *nd, void *mac)
+	if (!dev->nd->ops->close)
+		/* Can't kill the poll threads, so we can't do
+		 * anything safely. */
+		return 1;
```
Lets use negative numbers for errors, this is the convention we used so far.

pscollins replied:
Done

opurdila commented on tools/lkl/lib/virtio_net.c:
```diff
@@ -238,3 +238,42 @@ int lkl_netdev_add(struct lkl_netdev *nd, void *mac)
+			failure_count, registered_dev_idx);
+		return -1;
+	} else {
+		lkl_printf("Freed all %d netdevs successfully.\n",
```
I prefer no prints in case of success. In the test below we can check for failures instead.

pscollins replied:
Shrug, I still worry that it's easy to end up in a situation where the test doesn't catch a regression. But done.

pscollins commented:
In addition to addressing the comments, I fixed the build errors with DPDK and added some slightly more graceful error handling in `netdev_add` (so you may want to take a look at those spots, too).

stfairy commented on arch/lkl/include/uapi/asm/host_ops.h:
```diff
@@ -108,12 +112,14 @@ struct lkl_host_operations {
 int lkl_start_kernel(struct lkl_host_operations *lkl_ops,
-		     unsigned long mem_size,
-		     const char *cmd_line, ...);
+		int (*host_cleanup)(void),
```
Can we move this to lkl_host_operations?

stfairy commented on tools/lkl/lib/virtio_net_tap.c:
```diff
@@ -122,5 +207,23 @@ struct lkl_netdev *lkl_netdev_tap_create(const char *ifname)
+	if (lkl_netdev_tap_ops.eventfd) {
+		/* eventfd is supported by the host, all is well */
+		nd->dev.ops = &tap_net_ops_closeable;
+		nd->eventfd = lkl_netdev_tap_ops.eventfd(0, EFD_NONBLOCK | EFD_SEMAPHORE);
```
Keep line within 80 chars.

pscollins commented:
Review status: 0 of 16 files reviewed at latest revision, 3 unresolved discussions.

---

*[arch/lkl/include/uapi/asm/host_ops.h, line 121 \[r3\]](https://reviewable.io:443/reviews/lkl/linux/130#-KCw_xbGFr-hkfNFe6LL-r3-121:-KD9bs5vbF2iZn_nYbR1:651869726) ([raw file](https://github.com/lkl/linux/blob/bb881b86fba155cc84ee94ab9261a72ec89ec17a/arch/lkl/include/uapi/asm/host_ops.h#L121)):*
Done.

The slightly awkward thing is that 

---

*[tools/lkl/lib/virtio_net_tap.c, line 213 \[r3\]](https://reviewable.io:443/reviews/lkl/linux/130#-KCw_xbGFr-hkfNFe6LZ-r3-213:-KD9cV3yzE6WUkif0wNX:-499146009) ([raw file](https://github.com/lkl/linux/blob/bb881b86fba155cc84ee94ab9261a72ec89ec17a/tools/lkl/lib/virtio_net_tap.c#L213)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/130)*
<!-- Sent from Reviewable.io -->

liuyuan10 commented:
Review status: 0 of 12 files reviewed at latest revision, 6 unresolved discussions.

---

*[arch/lkl/include/uapi/asm/host_ops.h, line 64 \[r3\]](https://reviewable.io:443/reviews/lkl/linux/130#-KD9_9zu4p-x0uQ5MH8i:-KD9_9zyUgPqVdlxAjfR:1885024221) ([raw file](https://github.com/lkl/linux/blob/bb881b86fba155cc84ee94ab9261a72ec89ec17a/arch/lkl/include/uapi/asm/host_ops.h#L64)):*
I don't understand this comment. Could you elaborate?

---

*[tools/lkl/include/lkl.h, line 59 \[r3\]](https://reviewable.io:443/reviews/lkl/linux/130#-KD9_tL7WzlKB1au7B7-:-KD9_tL87kBxssPFuZ06:-1844927867) ([raw file](https://github.com/lkl/linux/blob/bb881b86fba155cc84ee94ab9261a72ec89ec17a/tools/lkl/include/lkl.h#L59)):*
Why do we need two declarations of lkl_netdev in this file?

And you say it's defined in virtio.h but why I see it in lkl_host.h?

---

*[tools/lkl/lib/virtio.h, line 65 \[r3\]](https://reviewable.io:443/reviews/lkl/linux/130#-KD9bRn-B0DRoxWk5AcO:-KD9bRn-B0DRoxWk5AcP:-1902957277) ([raw file](https://github.com/lkl/linux/blob/bb881b86fba155cc84ee94ab9261a72ec89ec17a/tools/lkl/lib/virtio.h#L65)):*
why is this in virtio.h? I think virtio knows nothing about tap? Is it better to have a special header file for tap? If you don't want a tap.h, is it possible to have it in virtio_net_tap.c but extern struct lkl_netdev_tap_ops lkl_netdev_tap_ops in hijack because it's now only used by hijack.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/130)*
<!-- Sent from Reviewable.io -->

pscollins commented:
Review status: 0 of 12 files reviewed at latest revision, 6 unresolved discussions.

---

*[arch/lkl/include/uapi/asm/host_ops.h, line 121 \[r3\]](https://reviewable.io:443/reviews/lkl/linux/130#-KCw_xbGFr-hkfNFe6LL-r3-121:-KD9fCDjzDRFM31T4owC:342693291) ([raw file](https://github.com/lkl/linux/blob/bb881b86fba155cc84ee94ab9261a72ec89ec17a/arch/lkl/include/uapi/asm/host_ops.h#L121)):*
Urg, re-pasted from github:

The slightly awkward thing is that it's a pain to get at lkl_remove_netdevs from inside the POSIX-host library, so I just set it up like virtio_devices is right now. In the future I think that maybe the lkl_host_ops struct should have a few subparts (e.g. host platform independent ops like printf and malloc, host platform dependent ops like mutex_lock, common data like virtio_devices, etc), but I think this works well enough for now. Let me know what you think.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/130)*
<!-- Sent from Reviewable.io -->

liuyuan10 commented on tools/lkl/include/lkl.h:
```diff
@@ -229,6 +233,17 @@ struct lkl_netdev {
 int lkl_netdev_add(struct lkl_netdev *nd, void *mac);
 
 /**
+* lkl_netdevs_remove - destroy all network devices
```
I think anyone that called lkl_netdev_add has to call lkl_netdevs_remove to avoid mem leak. Can you make it clear in the comment that one has to use "host_ops.host_clean = lkl_netdevs_remove"?

liuyuan10 replied:
@stfairy Do you see how this work with your facade? Exposing lkl_netdevs_remove in facade?

pscollins replied:
Hmmmm I think this was just not a good way to organize this. I did it this way because `lkl.h` isn't available from `arch/lkl` but between your comments and Xiao's I think it's probably better to just move those function prototypes into somewhere that's visible from `arch/lkl`.

pscollins commented:
Review status: 0 of 12 files reviewed at latest revision, 7 unresolved discussions.

---

*[arch/lkl/include/uapi/asm/host_ops.h, line 64 \[r3\]](https://reviewable.io:443/reviews/lkl/linux/130#-KD9_9zu4p-x0uQ5MH8i:-KD9fXZSqIqNYRrZFJTD:-1965821294) ([raw file](https://github.com/lkl/linux/blob/bb881b86fba155cc84ee94ab9261a72ec89ec17a/arch/lkl/include/uapi/asm/host_ops.h#L64)):*
`gettid` returns the result from `syscall(SYS_gettid)` (a Linux thread id), but `thread_create` returns the pthread thread id that comes out of `pthread_create`. So there doesn't need to be any relationship between the two values --- they could be the same, they could be different, etc.

---

*[tools/lkl/include/lkl.h, line 59 \[r3\]](https://reviewable.io:443/reviews/lkl/linux/130#-KD9_tL7WzlKB1au7B7-:-KD9gl8qWX1pJPyezMNW:1382027759) ([raw file](https://github.com/lkl/linux/blob/bb881b86fba155cc84ee94ab9261a72ec89ec17a/tools/lkl/include/lkl.h#L59)):*
Oh, whoops, that was from a bad merge while I was editing this commit, and the comment is outdated.

---

*[tools/lkl/lib/virtio.h, line 65 \[r3\]](https://reviewable.io:443/reviews/lkl/linux/130#-KD9bRn-B0DRoxWk5AcO:-KD9fPcbtdBo_QPuIXpV:138500941) ([raw file](https://github.com/lkl/linux/blob/bb881b86fba155cc84ee94ab9261a72ec89ec17a/tools/lkl/lib/virtio.h#L65)):*
I think this is going to be a little bit awkward no matter how it's done because `lkl_netdev_tap_ops` isn't needed without hijack, but I think it would be more confusing to have the `struct` defined twice in places that aren't visible to each other --- I'll make a `tap.h`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/130)*
<!-- Sent from Reviewable.io -->

pscollins commented:
@stfairy @liuyuan10 I've pushed up a new version that addresses your comments --- let me know what you think.

stfairy commented on arch/lkl/include/uapi/asm/host_ops.h:
```diff
@@ -113,7 +117,7 @@ struct lkl_host_operations {
-		     unsigned long mem_size,
-		     const char *cmd_line, ...);
+		unsigned long mem_size,
+		const char *cmd_line, ...);
```
fix whitespace

pscollins replied:
Done

stfairy commented on tools/lkl/include/lkl.h:
```diff
@@ -208,13 +208,11 @@ int lkl_if_set_ipv4(int ifindex, unsigned int addr, unsigned int netmask_len);
  */
 int lkl_set_ipv4_gateway(unsigned int addr);
 
+
```
remove extra blank line

pscollins replied:
Done

liuyuan10 commented on tools/lkl/include/lkl_host.h:
```diff
@@ -47,12 +47,21 @@ struct lkl_dev_blk_ops {
 	int (*request)(union lkl_disk disk, struct lkl_blk_req *req);
 };
 
+struct lkl_netdev {
```
Just curious, what's the motivation of moving it to lkl_host from lkl?

pscollins replied:
@liuyuan10 Because `lkl_netdev` needs to have an `lkl_thread_t` member, but `host_ops.h` isn't available in `lkl.h`. I guess it could be solved by adding an `include` to `lkl.h`, but I feel like there is a lot going on in `lkl.h` as it is, anyway. I think originally I needed to move it to `virtio.h` because I put a `virtio_net_dev` inside of an `lkl_netdev` and there was no other way to get at that type from there. As it is now I don't think there's a hugely compelling reason to do it one way or the other. Let me know if you think it should move back.

liuyuan10 replied:
I don't have preference either. LGTM

liuyuan10 commented:
LGTM

opurdila commented on arch/lkl/kernel/setup.c:
```diff
@@ -139,6 +141,10 @@ long lkl_sys_halt(void)
 	lkl_ops->sem_free(init_sem);
 
 	free_initial_syscall_thread();
+	if (lkl_netdevs_remove() == 0)
```
I think we need some sort of cleanup infrastructure, where host can register cleanup routines. This would avoid propagation of specific host functions in the kernel code which IMO would be cleaner.

However, I think we can merge this PR as it is and I will log an issue to track that separately.
